swap/chain/backend.go

```diff
@@ -0,0 +1,78 @@
+package chain
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind/backends"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// Backend is the minimum amount of functionality required by the underlying ethereum backend
+type Backend interface {
+	bind.ContractBackend
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+	TransactionByHash(ctx context.Context, txHash common.Hash) (*types.Transaction, bool, error)
+}
+
+// TestBackend is the backend to use for tests with a simulated backend
+type TestBackend struct {
```

Comment:
what's the right thing to do if I need the same test helpers (e.g. `TestBackend`) in tests in multiple packages? It seems functions from `_test.go` files are never exported even during testing.

Reply:
well you can reuse it the same way you reuse test helpers inside `common_test.go`, i think this is where it should reside. if you want to have a real solid mock that is used across several packages (i'm not sure if i see a necessity for this) you should have a mock package and inject that when necessary according to a clean interface between the packages.

Reply:
I agree that the best approach is to have a clean mock in a dedicated package that is implementing the interface. That way, the mock package will be imported only by tests and not compiled in binary. And also reused between multiple dependents of the same interface.

Reply:
I moved the `TestBackend` to a separate `mock` package. The main reason for this move is that we are moving the transaction sending and monitoring logic into a separate component. As this will be independent of the rest of `swap`, the idea was to have this in a separate package (`swap/chain`) which also would require the `Backend` interface and the `TestBackend` mock.